src/test/mina_automation/executor.ml

```diff
@@ -75,49 +73,111 @@ module Make (P : AppPaths) = struct
-    Util.run_cmd_exn ?env "." full_path args
-
-  let run_from_dune ~(args : string list) ?env () =
-    log_executed_command PathFinder.app_name_under_dune ;
-    Util.run_cmd_exn ?env "." "dune"
-      ([ "exec"; PathFinder.app_name_under_dune; "--" ] @ args)
+    Util.create_process_exn ?env "." full_path args ()
 
-  let run_from_local ~(args : string list) ?env () =
-    log_executed_command PathFinder.built_name ;
-    Util.run_cmd_exn ?env "." PathFinder.built_name args
+  let run_from_debian ?(prefix = "") ~(args : string list) ?env
+      ?(ignore_failure = false) () =
+    let%bind output =
+      run_from_debian_in_background ~prefix ~args ?env ()
+      >>= Process.collect_output_and_wait
+    in
+    if ignore_failure then return output.stdout
+    else output_or_hard_error ~prog:PathFinder.Paths.official_name ~args output
+
+  let run_from_dune_in_background ~args ?env () =
+    log_executed_command PathFinder.Paths.dune_name ;
+    Util.create_process_exn ?env "." "dune"
+      ([ "exec"; PathFinder.Paths.dune_name; "--" ] @ args)
+      ()
+
+  let run_from_dune ~(args : string list) ?env ?(ignore_failure = false) () =
+    let%bind output =
+      run_from_dune_in_background ~args ?env ()
+      >>= Process.collect_output_and_wait
+    in
+    if ignore_failure then return output.stdout
+    else output_or_hard_error ~prog:PathFinder.Paths.dune_name ~args output
 
-  let run t ~(args : string list) ?env () =
+  let run t ~(args : string list) ?env ?ignore_failure () =
     let open Deferred.Let_syntax in
     match t with
     | AutoDetect -> (
         match%bind Sys.file_exists PathFinder.built_name with
         | `Yes ->
-            run_from_local ~args ?env ()
+            run_from_local ~args ?env ?ignore_failure ()
         | _ -> (
             match%bind
               Deferred.List.find_map
-                ~f:(PathFinder.exists_at_path PathFinder.official_name)
+                ~f:(PathFinder.exists_at_path PathFinder.Paths.official_name)
                 PathFinder.paths
             with
             | Some prefix ->
-                run_from_debian ~prefix ~args ?env ()
+                run_from_debian ~prefix ~args ?env ?ignore_failure ()
             | _ ->
-                run_from_dune ~args ?env () ) )
+                run_from_dune ~args ?env ?ignore_failure () ) )
     | Dune ->
-        run_from_dune ~args ?env ()
+        run_from_dune ~args ?env ?ignore_failure ()
     | Debian ->
-        run_from_debian ~args ?env ()
+        run_from_debian ~args ?env ?ignore_failure ()
     | Local ->
-        run_from_local ~args ?env ()
+        run_from_local ~args ?env ?ignore_failure ()
     | Docker ctx ->
         let docker = Docker.Client.default in
         let cmd = [ P.official_name ] @ args in
         Docker.Client.run_cmd_in_image docker ~image:ctx.image ~cmd
           ~workdir:ctx.workdir ~volume:ctx.volume ~network:ctx.network
+
+  let run_in_background t ~(args : string list) ?env () =
```

Comment:
This is basically copy-paste of `run`, so I assume it would also be possible to factor out a common pattern.